### PR TITLE
ENH: Add get version for Intel C Compiler and tweaked options

### PR DESCRIPTION
**Pull request by oleksandr-pavlyk**
@eric-wieser 

This PR adds changes to `distutils/intelccompiler.py` and `distutils/fcompiler/intel.py` from fork [IntelPython/numpy](http://github.com/IntelPython/numpy).

Changes add a routine to get a version of Intel C Compiler used, allowing to use`-qopenmp` for recent version, while using `-fopenmp` for much older ones.

However, I commented out use `openmp` in the compiler string to avoid overlinking of NumPy itself, which does not use OpenMP at all. 

Those native extensions which require OpenMP should explicitly set the compiler switch.

### CODE REVIEW

**Comment by charris** [comment]
Test failure is unrelated.

File: numpy/distutils/fcompiler/intel.py [hunk 1]
```diff
@@ -27,6 +27,17 @@ def runtime_library_dir_option(self, dir):
         return '-Wl,-rpath=%s' % dir
 
 
+    def get_version(self):
+        if platform.system() == 'Windows':
+            version_cmd = "icl -dummy"
+            status, output = exec_command(version_cmd, use_tee=0)
```
Comment (eric-wieser) [inline numpy/distutils/fcompiler/intel.py:33]: I believe `exec_command` is now deprecated, and this should either use self.spawn or subprocess.

Comment (charris) [reply numpy/distutils/fcompiler/intel.py:33]: @oleksandr-pavlyk exec_command was deprecated in 1.17.

Comment (charris) [reply numpy/distutils/fcompiler/intel.py:33]: @oleksandr-pavlyk Ping.

Comment (oleksandr-pavlyk) [reply numpy/distutils/fcompiler/intel.py]: I pushed the change to use `subprocess`.

File: numpy/distutils/fcompiler/intel.py [hunk 2]
```diff
@@ -27,6 +28,19 @@ def runtime_library_dir_option(self, dir):
...
+        if platform.system() == 'Windows':
+            version_cmd = ["icl", "-dummy"]
+            completed_process = subprocess.run(version_cmd, stdout=PIPE, stderr=STDOUT)
+            status, output = completed_process.returncode, completed_process.stdout
+            version = re.search(r'Version\s*([\d.]+)', output).group(1)
+        else:
+            version_cmd = ["icc", "-dumpversion"]
+            completed_process = subprocess.run(version_cmd, stdout=PIPE, stderr=STDOUT)
+            status, output = completed_process.returncode, completed_process.stdout
+        return version
```
Comment (eric-wieser) [inline numpy/distutils/fcompiler/intel.py:41]: This gives `NameError` on linux

**Review by eric-wieser — DISMISSED** [review]
This change completely undermines the rest of the version infrastructure in this file. `FCompiler` extends `CCompiler`, and our `CCompiler.get_version()` is implemented as:
https://github.com/numpy/numpy/blob/43683b3256a86659f230dcadbcde1f8020398bfa/numpy/distutils/ccompiler.py#L588-L655

In short:

* `self.executables['version_cmd']` contains the appropriate command line
* Either:
   * `self.version_match` contains a suitable function for matching.
   * `self.version_pattern` contains a suitable regex.

The rest of this file is already taking care to set this up. If it's not working, we should fix this setup, not completely ignore it.

File: numpy/distutils/fcompiler/intel.py (hunk 2 above)
Comment (oleksandr-pavlyk) [reply numpy/distutils/fcompiler/intel.py]: Oops. Serves me well to push without testing.

I squashed and pushed the locally tested change.

**Comment by oleksandr-pavlyk** [comment]
@eric-wieser Thank you for the feedback. I have pushed changes which set `self.executables['version_cmd']` and `version_pattern`. 

I tested by building locally, but that did not exercise the Intel(R) Fortran compiler part. Hopefully CI, or actual review can cover that.

File: numpy/distutils/fcompiler/intel.py [hunk 3]
```diff
@@ -15,6 +15,17 @@ def intel_version_match(type):
...
+            self.executables['version_cmd'] = ["icl" , "-dummy"]
+            self.version_pattern = r'Version\s*(?"version"[\d.]+)'
+        else:
+            self.executables['version_cmd'] = ["icc", "-dumpversion"]
+            self.version_pattern = r'^(?"version"[\d.]$)'
+        
+    
     def update_executables(self):
         f = dummy_fortran_file()
         self.executables['version_cmd'] = ['<F77>', '-FI', '-V', '-c',
```
Comment (eric-wieser) [inline numpy/distutils/fcompiler/intel.py:31]: `self.executables['version_cmd']` is already set here!

File: numpy/distutils/fcompiler/intel.py [hunk 4]
```diff
@@ -15,6 +15,17 @@ def intel_version_match(type):
...
 class BaseIntelFCompiler(FCompiler):
+    def __init__(self, *args, **kwargs):
+        FCompiler.__init__(self, *args, **kwargs)
+        # specify version discovery bits
+        if platform.system() == "Windows":
+            self.executables['version_cmd'] = ["icl" , "-dummy"]
+            self.version_pattern = r'Version\s*(?"version"[\d.]+)'
+        else:
+            self.executables['version_cmd'] = ["icc", "-dumpversion"]
+            self.version_pattern = r'^(?"version"[\d.]$)'
```
Comment (eric-wieser) [inline numpy/distutils/fcompiler/intel.py:26]: Similarly, the subclasses already define `version_match`, which takes precedence over `version_pattern`.

**Comment by oleksandr-pavlyk** [comment]
Is this good, more changes needed, or should be closed?

File: numpy/distutils/intelccompiler.py [hunk 5]
```diff
@@ -16,10 +17,19 @@ class IntelCCompiler(UnixCCompiler):
...
+
+        # Using -qopenmp leads to overlinking if C code does not use OpenMP features
+        # like in the case of NumPy. If an extension requires -qopenmp, it should be
+        # explicitly set for that extension
         v = self.get_version()
-        mpopt = 'openmp' if v and v < '15' else 'qopenmp'
-        self.cc_exe = ('icc -fPIC -fp-model strict -O3 '
-                       '-fomit-frame-pointer -{}').format(mpopt)
+        mpopt = '-openmp' if v and v < '15' else '-qopenmp'
+        self.cc_exe = 'icc -fPIC -fp-model strict -O3 -fomit-frame-pointer ' # + mpopt
```
Comment (eric-wieser) [inline numpy/distutils/intelccompiler.py:32]:
```suggestion
        self.cc_exe = 'icc -fPIC -fp-model strict -O3 -fomit-frame-pointer '
```
If the code is being removed, lets remove it completely.

File: numpy/distutils/intelccompiler.py [hunk 6]
```diff
@@ -56,10 +66,19 @@ class IntelEM64TCCompiler(UnixCCompiler):
...
+
+        # Using -qopenmp leads to overlinking if C code does not use OpenMP features
+        # like in the case of NumPy. If an extension requires -qopenmp, it should be
+        # explicitly set for that extension
         v = self.get_version()
-        mpopt = 'openmp' if v and v < '15' else 'qopenmp'
-        self.cc_exe = ('icc -m64 -fPIC -fp-model strict -O3 '
-                       '-fomit-frame-pointer -{}').format(mpopt)
+        mpopt = '-openmp' if v and v < '15' else '-qopenmp'
+        self.cc_exe = 'icc -m64 -fPIC -fp-model strict -O3 -fomit-frame-pointer' # + mpopt
```
Comment (eric-wieser) [inline numpy/distutils/intelccompiler.py:81]: It puzzles me that we decide to use `icl` on windows above, yet are fine using `icc` here unconditionally. Is this correct?

**Review by eric-wieser — COMMENTED** [review]
Looks like the version stuff I objected to from the previous revision is gone, thanks

**Comment by InessaPawson** [comment]
Given numpy.distutils has been officially deprecated, should we close this PR?

**Pull request closed without merging**